File: core/src/Adjustment.cpp
#include "cmaclib/Adjustment.h"

#include <cmaclib/Result.h>
#include <utility>

using namespace CmacLib;

void Adjustment::SetWeightChanges(std::vector<std::vector<double>>&& dw)
{
	this->weightChanges = ::std::move(dw);
}

Adjustment::Adjustment()
	: weightChanges(std::vector<std::vector<double>>()){}

Adjustment::~Adjustment(){}

const std::vector<std::vector<double>>& Adjustment::GetWeightChanges()
{
	return this->weightChanges;
}

/// @brief Get the amount the each active weight changed.
/// @param pAdjustment The pointer to the adjustment instance.
/// @return Weight change dW.
/// (Remember to delete the CDoubleArray data pointer.)
CDoubleArrayArray CCmacLibIAdjustmentGetWeightChanges(
	CCmacLibIAdjustment* pAdjustment
) {
	/// @brief The container for the active weights 2-D array.
	CDoubleArrayArray matrix;
	matrix.size = pAdjustment->GetWeightChanges().size();
	matrix.data = new CDoubleArray[matrix.size];

	for (unsigned long i = 0; i < matrix.size; i++) {
		matrix.data[i].data = (double*) pAdjustment->GetWeightChanges().data()->data();
	}

	return matrix;
}

/// @brief Indicates if an operation was successful.
/// @param pAdjustment The pointer to the adjustment instance.
/// @return Success.
cmaclib_bool_t CCmacLibIAdjustmentIsSuccessful(
	CCmacLibIAdjustment* pAdjustment
) {
	return pAdjustment->IsSuccessful();
}

/// @brief If not successful, a failure message is provided.
/// @param pAdjustment The pointer to the adjustment instance.
const char* CCmacLibIAdjustmentGetMessage(
	CCmacLibIAdjustment* pAdjustment
) {
	return pAdjustment->GetMessage().c_str();
}

/// @brief Destructor.
/// @param pAdjustment The pointer to the adjustment instance.
void CCmacLibIAdjustmentDestroy(
	CCmacLibIAdjustment* pAdjustment
) {
	if (pAdjustment) delete pAdjustment;
}

File: wrappers/julia/src/adjustment.jl
#!/usr/bin/julia

export IAdjustment
"""
Contains weight adjustment data.
"""
struct IAdjustment end

export IAdjustmentGetWeightChanges
"""
Get the amount the each active weight changed.
## Arguments
- `pInst::Ptr{IAdjustment}`: The pointer to the adjustment instance.
## Returns
Weight change dW.
"""
function IAdjustmentGetWeightChanges(pInst::Ptr{IAdjustment})::Vector{Vector{Cdouble}}
    c_data = ccall((:CCmacLibIAdjustmentGetWeightChanges, cmaclib_core),
        CDoubleArrayArray, (Ptr{IPrediction}, ), pInst)

    # The container for the active weights to be returned.
    matrix::Vector{Vector{Cdouble}} = []
    c_matrix::Vector{CDoubleArray} = unsafe_wrap(Vector{CDoubleArray},
        c_data.data, c_data.size)

    for col_vec_wrapped::CDoubleArray in c_matrix
        col_vec::Vector{Cdouble} = unsafe_wrap(
            Vector{Cdouble}, col_vec_wrapped.data, col_vec_wrapped.size
        )
        push!(matrix, col_vec)
    end

    # Properly Dispose c_data.
    ccall((:CDoubleArrayArrayDestroy, cmaclib_core),
        Cvoid, (CDoubleArrayArray, ), c_data)

    return matrix
end

export IAdjustmentIsSuccessful
"""
Indicates if an operation was successful.
## Arguments
- `pInst::Ptr{IAdjustment}`: The pointer to the adjustment instance.
## Returns
Success
"""
function IAdjustmentIsSuccessful(pInst::Ptr{IAdjustment})::Bool
    return ccall((:CCmacLibIAdjustmentIsSuccessful, cmaclib_core),
        Bool, (Ptr{IAdjustment}, ), pInst)
end

export IAdjustmentGetMessage
"""
If not successful, a failure message is provided.
## Arguments
- `pInst::Ptr{IAdjustment}`: The pointer to the adjustment instance.
## Returns
The failure message in C string.
"""
function IAdjustmentGetMessage(pInst::Ptr{IAdjustment})::Cstring
    return ccall((:CCmacLibIAdjustmentGetMessage, cmaclib_core),
        Cstring, (Ptr{IAdjustment}, ), pInst)
end

export IAdjustmentDestroy
"""
Destructor.

## Arguments
- `pInst::Ptr{IAdjustment}`: The pointer to the adjustment instance.
"""
function IAdjustmentDestroy(pInst::Ptr{IAdjustment})::Nothing
    ccall((:CCmacLibIAdjustmentDestroy, cmaclib_core),
        Cvoid, (Ptr{IAdjustment}, ), pInst)
end